bookkeeper-common/src/main/java/org/apache/bookkeeper/common/collections/BatchedArrayBlockingQueue.java

```diff
@@ -0,0 +1,409 @@
+            }
+
+            return dequeueOne();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public int remainingCapacity() {
+        return capacity - size;
+    }
+
+    @Override
+    public int drainTo(Collection<? super T> c) {
+        return drainTo(c, capacity);
+    }
+
+    @Override
+    public int drainTo(Collection<? super T> c, int maxElements) {
+        lock.lock();
+        try {
+            int toDrain = Math.min(size, maxElements);
+
+            int consumerIdx = this.consumerIdx;
+            for (int i = 0; i < toDrain; i++) {
+                T item = data[consumerIdx];
+                data[consumerIdx] = null;
+                c.add(item);
+
+                if (++consumerIdx == capacity) {
+                    consumerIdx = 0;
+                }
+            }
+
+            this.consumerIdx = consumerIdx;
+            if (size == capacity) {
+                notFull.signalAll();
+            }
+
+            size -= toDrain;
+            return toDrain;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public int takeAll(T[] array) throws InterruptedException {
+        return internalTakeAll(array, true, 0, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public int pollAll(T[] array, long timeout, TimeUnit unit) throws InterruptedException {
+        return internalTakeAll(array, false, timeout, unit);
+    }
+
+    private int internalTakeAll(T[] array, boolean waitForever, long timeout, TimeUnit unit)
+            throws InterruptedException {
+        lock.lockInterruptibly();
+        try {
+            while (size == 0) {
+                if (waitForever) {
+                    notEmpty.await();
+                } else {
+                    if (!notEmpty.await(timeout, unit)) {
+                        return 0;
+                    }
+                }
+            }
+
+            int toDrain = Math.min(size, array.length);
+
+            int consumerIdx = this.consumerIdx;
+
+            // First span
+            int firstSpan = Math.min(toDrain, capacity - consumerIdx);
+            System.arraycopy(data, consumerIdx, array, 0, firstSpan);
+            Arrays.fill(data, consumerIdx, consumerIdx + firstSpan, null);
+            consumerIdx += firstSpan;
+
+            int secondSpan = toDrain - firstSpan;
+            if (secondSpan > 0) {
+                System.arraycopy(data, 0, array, firstSpan, secondSpan);
+                Arrays.fill(data, 0, secondSpan, null);
+                consumerIdx = secondSpan;
+            }
+
+            if (consumerIdx == capacity) {
+                consumerIdx = 0;
+            }
+            this.consumerIdx = consumerIdx;
+            if (size == capacity) {
+                notFull.signalAll();
+            }
+
+            size -= toDrain;
+            return toDrain;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void clear() {
```

Comment:
And the clear() may not reach the expectation, it may clear the queue.

```
        new Thread(()-> {
            Integer[] integers = new Integer[] {1, 2, 3, 4, 5, 6, 7, 8, 9,10,11,12,13,14,15,16,17,18,19,20};
            try {
                queue.putAll(integers, 0, 20);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }).start();
        
        
        queue.clear();
        //After clear, the queue is not empty.
        System.out.println(queue.size());
```